package.json

```diff
@@ -2,11 +2,16 @@
   "author": "Stefan Goessner",
   "name": "jsonpath-plus",
   "version": "5.0.2",
-  "type": "commonjs",
```

Comment:
Please keep this. It helps make explicit how the behavior works. See https://nodejs.org/api/packages.html for the standard that webpack and Rollup plugins are or ought to be (and likely will be if they aren't already) following.

Reply:
> Please keep this. It helps make explicit how the behavior works. See https://nodejs.org/api/packages.html for the standard that webpack and Rollup plugins are or ought to be (and likely will be if they aren't already) following.

Sorry, but with `"type": "commonjs"` babel/webpack tell me:
```
ERROR in ./node_modules/jsonpath-plus/dist/index-browser-esm.js 1061:0
Module parse failed: 'import' and 'export' may appear only with 'sourceType: module' (1061:0)
```

Reply:
Hmm... I've added a comment at https://github.com/webpack/webpack/issues/11597#issuecomment-772579321 , as it looks to me that bundlers should not be using `type` for interpretation of non-Node fields. There is a [workaround](https://github.com/webpack/webpack/issues/11597#issuecomment-705114717) there that may work for now without our changing the code, as I really think we should be able to use ".js" with browser ESM while using `type: commonjs` for Node. If we don't hear back in some time, let's file a new Webpack issue.

Reply:
And when revisiting this, Node recommends using `default` over `browser` (that probably makes sense too since other environments might support our own "vm" but not Node's).

Reply:
> There is a [workaround](https://github.com/webpack/webpack/issues/11597#issuecomment-705114717) there that may work for now without our changing the code

If by that work-around you mean adding something like:
```
  module: {
    rules: [
      {
        test: /browser-esm\.js$/,
        include: /node_modules\/jsonpath-plus/,
        type: "javascript/auto"
      },
      …
    ]
  }
```
to my webpack config, then: no, that doesn't work. Webpack 5 still loads the node version.

Reply:
Even a desparate attempt such as:
```
  module: {
    resolve: {
      alias: {
        "jsonpath-plus": "jsonpath-plus/dist/index-browser-esm.js"
      }
    }
  } 
```
fails with a Webpack 5 error:
```  
[webpack-cli] ModuleNotFoundError: Module not found: Error: Package path ./dist/index-browser-esm.js is not exported from package …/node_modules/jsonpath-plus (see exports field in …/node_modules/jsonpath-plus/package.json)
```
The only viable solution is to change the exports in `package.json`, it would seem.

Reply:
> And when revisiting this, Node recommends using `default` over `browser`

I've added `default`, but since [this condition should always come last](https://nodejs.org/api/packages.html#packages_conditional_exports), I've merely added it at the end; `browser` needs to stay before `import`/`require`.

Reply:
Ok. I presume that dropping the `browser` in favor of just `default` doesn't help as it may fall back to it eventually anyways if it is not picking up the other types.

There should be no difference in having `type: 'commonjs'` and having no `type`, as the former is the default per https://nodejs.org/api/packages.html#packages_type .

We might look to switch to `type: 'module'` and then have our Rollup build ".cjs" files for Node, as this makes the intent clear (and also is even more accurate given our source files being ESM). Node should understand `.cjs` without problem. However, I was not able to get Mocha to work, apparently because of https://github.com/mochajs/mocha/issues/3060#issuecomment-506511488 . While `--parallel` showed some (imperfect) results, this is not compatible with the `--delay` option our tests currently need.

Reply:
> Ok. I presume that dropping the `browser` in favor of just `default` doesn't help as it may fall back to it eventually anyways if it is not picking up the other types.

Since the export conditions are processed in order, specifying:
```

Reply:
> There should be no difference in having `type: 'commonjs'` and having no `type`, as the former is the default per https://nodejs.org/api/packages.html#packages_type .

That may be so, but: https://github.com/JSONPath-Plus/JSONPath/pull/146#discussion_r569460883

Reply:
> We might look to switch to `type: 'module'` and then have our Rollup build ".cjs" files for Node, as this makes the intent clear (and also is even more accurate given our source files being ESM).

Not sure if the advice in https://nodejs.org/api/packages.html#packages_dual_commonjs_es_module_packages would help you there. In any case, it would be nice if there would at least be _some_ release that would allow us to use the proper browser version of the code.

Reply:
Have you tried again with `type: 'commonjs'` added back? Is the issue hat it triggers even if a condition like `browser` is met earlier?

Reply:
I went ahead and filed https://github.com/webpack/webpack/issues/12598 as I do recall you had this before. Will see about a response.

Reply:
Given our earlier discussion of this, I wanted to let you know (and anyone watching this) that I am planning to release as a true ESM module by using `type: 'module'` (otherwise our default ".js" is not understood in some environments as being a genuine module, e.g., Mocha 9 fails completely, and I think it has only been working in modules up to now because ESM can import CJS as we provide).

Would you mind testing out the `esm` branch?

Reply:
@brettz9 Where do I find this `esm` branch?

Reply:
Oh, sorry, https://github.com/brettz9/JSONPath/tree/esm , i.e., cloning https://github.com/brettz9/JSONPath#esm

Reply:
@brettz9 Appears to work fine; apologies for the delay.